Fix reading Series with read_hdf

frexvahi opened the pull request:
- [x] closes #16583
 - [x] tests added / passed
 - [x] passes ``git diff upstream/master --name-only -- '*.py' | flake8 --diff``
 - [x] whatsnew entry

frexvahi committed:
Added test to reproduce issue #16583

jreback commented on pandas/io/pytables.py:
```diff
@@ -313,10 +313,14 @@ def read_hdf(path_or_buf, key=None, **kwargs):
         """
 
-    if kwargs.get('mode', 'a') not in ['r', 'r+', 'a']:
+    # Ignore any supplied mode, we only need to read data
```
this is the not addressing the issue. you are bypassing the mode.

frexvahi replied:
Please explain to me why the read_hdf function would ever want to use any mode other than 'r'. It's not like the returned object is linked to the file by an mmap - the file is opened, data are read, the file is closed. At no point is there any need to write to the file.

jreback replied:
I am not averse to this change, but its different. There is an actual issue here where a lower level function is mistakenly passed the mode arg. Pls solve that issue first.

jreback replied:
This is actually tricky, because IIRC HDF5 has some locking issues when some files are read-only and others are append, which is why the default is append (in a multi-open e.g. process/thread context).

frexvahi replied:
OK, I fixed it a different way and split the change to the default mode into a separate commit so you can accept it or reject it.

frexvahi replied:
I just had another thought - even if we want to open in read-write mode because of the locking issues, the default should be 'r+' rather than 'a' so that trying to read a non-existent file gives a FileNotFoundError.

frexvahi replied:
cancel that, there is already code to check for existence and raise FileNotFoundError

jreback commented on pandas/io/pytables.py:
```diff
@@ -348,6 +348,10 @@ def read_hdf(path_or_buf, key=None, **kwargs):
         # so delegate to the iterator
         auto_close = True
 
+    # Remove kwargs which are only relevant for the HDFStore (see GH#16583)
```
I would like a narrower change, better to simply name the args here, then they don't need to be passed in.

frexvahi replied:
As in, add `mode='r'` to the signature of `read_hdf`?

frexvahi committed:
Fix #16583 by adding an explicit `mode` argument to `read_hdf`

kwargs which are meant for the opening of the HDFStore should be filtered out
before passing the remaining kwargs to the `select` function to load the data.

jreback commented on pandas/io/pytables.py:
```diff
@@ -282,21 +282,23 @@ def to_hdf(path_or_buf, key, value, mode=None, complevel=None, complib=None,
         ----------
-        path_or_buf : path (string), buffer, or path object (pathlib.Path or
-            py._path.local.LocalPath) to read from
+        path_or_buf : path (string), buffer, open pd.HDFStore, or path object
```
use comma to separate except last clause

jreback commented on pandas/io/pytables.py:
```diff
@@ -282,21 +282,23 @@ def to_hdf(path_or_buf, key, value, mode=None, complevel=None, complib=None,
 
         key : group identifier in the store. Can be omitted if the HDF file
             contains a single pandas object.
+        mode : Mode to use when opening the file (ignored if path_or_buf is
```
string, {'r', 'r+', 'a', 'w'}, default 'r'

then you can say ignored on the next line

jreback commented on pandas/tests/io/test_pytables.py:
```diff
@@ -5191,6 +5191,14 @@ def test_query_compare_column_type(self):
                     tm.assert_frame_equal(expected, result)
 
+    def test_read_hdf_series_fixed(self):
+        # GH 16583
```
add a 1-liner on what the issue you are testing is

frexvahi replied:
I added a description but GitHub does not seem to have detected this

jreback reviewed (changes requested):
lgtm. pls add a note in 20.3 IO bug fixes. some comments. ping on green.

frexvahi committed:
Noted fix for #16583 in WhatsNew

jreback commented on pandas/tests/io/test_pytables.py:
```diff
@@ -5247,6 +5247,16 @@ def test_query_compare_column_type(self):
+        # GH 16583
+        # Tests that reading a Series saved to an HDF file in 'fixed' format
+        # still works if a mode='r' argument is supplied
+        series = tm.makeFloatSeries()
```
can you do this also for a ``format='table'`` (and rename the test slightly). It might duplicate an existing test, but that's ok.

ping when pushed / green. lgtm merge otherwise.

frexvahi replied:
Done

frexvahi commented:
The Travis CI tests seem to be failing because of a MemoryError in PyArrow, I'm not sure what that could have to do with this change.

jreback commented:
@frexvahi thanks for the report. Yes something happened on Travis with this import. This PR lgtm. Will merge soon. thanks.

jreback commented:
thanks!
